### DISCUSSION TITLE
How does this project compare to other popular solution like gun.js and orbitdb?

### DISCUSSION THREAD

**Discussion by sanity** [freenet/freenet-core#111] [Q&A] [answered]
(copied from Discord discussion for posterity)

**Comment by sanity** [comment]
Not super-familiar with gun.js and orbitdb, but I think the AWS analogy might be that they are analogous to replacing dynamodb while locutus is more akin to replacing the entire cloud infrastructure with something  decentralized.

It's also not quite clear whether gun.js and orbitdb provide a single unified database shared globally, or more private db clusters that are internally decentralized. Locutus is a unified global database with extremely flexible access control mediated by webassembly cryptographic contracts.

Those are my initial impressions but a proper comparison would need more research/thought.

**Comment by ple1n** [comment, marked as answer]
Many decentralized protocols could be seen as a key-value store with certain rules. Each peer has its own state, and the states get propagated when something changes. IPFS is a key-value store where a CID is associated with its providers and value. IPNS and Zeronet use publickey as the key, and for ZeroNet part of the value is treated as rules for updates to the state, while IPNS only validates the signer. In IPFS only the key contributes to the updating condition, which requires the value's hash to match the key.

By using a smart contract as the key, it means to use the hash of a contract as the key and treat the contract as the updating condition. This is generalization thanks to wasm. IPFS's contract is hash function so it's static. Locutus is dynamic and more generalized than zeronet which uses some fixed fields in site metadata to specify the updating condition which is not elegant.
